File: lib/faye/engine/memory.rb
module Faye
  module Engine
    
    class Memory < Base
      def initialize(options = {})
        @clients   = {}
        @channels  = {}
        @namespace = Namespace.new
        super
      end
      
      def create_client_id(&callback)
        client_id = @namespace.generate
        @clients[client_id] = Set.new
        ping(client_id)
        callback.call(client_id)
      end
      
      def client_exists?(client_id, &callback)
        callback.call(@clients.has_key?(client_id))
      end
      
      def ping(client_id)
        return unless Numeric === timeout
        remove_timeout(client_id)
        add_timeout(client_id, 2*timeout) { disconnect(client_id) }
      end
      
      def subscribe(client_id, channel)
        @clients[client_id] ||= Set.new
        @channels[channel] ||= Set.new
        @clients[client_id].add(channel)
        @channels[channel].add(client_id)
      end
      
      def unsubscribe(client_id, channel)
        @clients[client_id].delete(channel) if @clients.has_key?(client_id)
        @channels[channel].delete(client_id) if @channels.has_key?(channel)
      end
      
      def distribute(message)
        return if message['error']
        Channel.expand(message['channel']).each do |channel|
          next unless clients = @channels[channel]
          clients.each { |client_id| announce(client_id, message) }
        end
      end
      
      def disconnect(client_id)
        return unless @clients.has_key?(client_id)
        @clients[client_id].each { |channel| unsubscribe(client_id, channel) }
        @clients.delete(client_id)
      end
    end
   
  end
end

File: javascript/engine/memory.js
Faye.Engine.Memory = Faye.Class(Faye.Engine.Base, {
  initialize: function(options) {
    this._clients   = {};
    this._channels  = {};
    this._namespace = new Faye.Namespace();
    
    Faye.Engine.Base.prototype.initialize.call(this, options);
  },
  
  createClientId: function(callback, scope) {
    var clientId = this._namespace.generate();
    this._clients[clientId] = new Faye.Set();
    this.ping(clientId);
    callback.call(scope, clientId);
  },
  
  clientExists: function(clientId, callback, scope) {
    callback.call(scope, this._clients.hasOwnProperty(clientId));
  },
  
  ping: function(clientId) {
    var timeout = this._options.timeout;
    if (typeof timeout !== 'number') return;
    this.removeTimeout(clientId);
    this.addTimeout(clientId, 2 * timeout, function() {
      this.disconnect(clientId);
    }, this);
  },
  
  subscribe: function(clientId, channel) {
    var clients = this._clients, channels = this._channels;
    clients[clientId] = clients[clientId] || new Faye.Set();
    channels[channel] = channels[channel] || new Faye.Set();
    clients[clientId].add(channel);
    channels[channel].add(clientId);
  },
  
  unsubscribe: function(clientId, channel) {
    var clients = this._clients, channels = this._channels;
    if (clients.hasOwnProperty(clientId)) clients[clientId].remove(channel);
    if (channels.hasOwnProperty(channel)) channels[channel].remove(clientId);
  },
  
  distribute: function(message) {
    if (message.error) return;
    var channels = Faye.Channel.expand(message.channel);
    Faye.each(channels, function(channel) {
      var clients = this._channels[channel];
      if (!clients) return;
      clients.forEach(function(clientId) {
        this.announce(clientId, message);
      }, this);
    }, this);
  },
  
  disconnect: function(clientId) {
    var clients = this._clients;
    if (!clients.hasOwnProperty(clientId)) return;
    clients[clientId].forEach(function(channel) {
      this.unsubscribe(clientId, channel);
    }, this);
    delete clients[clientId];
  }
});

